client/file-browser/get-files.ts

```diff
@@ -13,34 +13,36 @@ function splitExtension(filename: string): [base: string, ext: string] {
 }
 
 export default async function readFolder(folderPath: string): Promise<FileBrowserEntry[]> {
-  try {
-    const entries = (await ipcRenderer.invoke('fsReadDir', folderPath, { withFileTypes: true }))!
-    return await Promise.all(
-      entries.map(async entry => {
-        const isFolder = entry.isDirectory
-        const [name, extension] = isFolder ? [entry.name, ''] : splitExtension(entry.name)
-        const stats = await ipcRenderer.invoke('fsStat', folderPath + '\\' + entry.name)
-        const path = folderPath + '\\' + entry.name
-
-        return isFolder
-          ? {
-              type: FileBrowserEntryType.Folder,
-              name,
-              path,
-            }
-          : {
-              type: FileBrowserEntryType.File,
-              name,
-              path,
-              extension: extension.toLowerCase(),
-              date: stats?.mtime ?? new Date(0),
-            }
-      }),
-    )
-  } catch (err) {
-    if ((err as any).code === 'ENOENT') {
-      throw new Error('Filepath ' + folderPath + " doesn't exist")
-    }
-    throw err
+  const readDirResult = (await ipcRenderer.invoke('fsReadDir', folderPath, {
+    withFileTypes: true,
+  }))!
+  if (readDirResult.error && readDirResult.code !== FsErrorCode.ENOENT) {
```

Comment:
You need to handle all errors that `fsReadDir` might return in its results here, not just specific ones. This will also allow you not needing to check for the errors below when using the entries.

Also, this still needs to be inside a try-catch block, to handle all the regular errors that we're not special-handling.

Reply:
This line is checking for non-ENOENT errors, and the next line handles the ENOENT error case, so all errors are covered. It is a little odd though since error handling is mixed between throwing and error codes. I wonder If I should go ahead and add all the error codes to the enum? That'll lead to a lot of boiler plate for what will end up being a small number of special cases.

Currently, every error other than ENOENT is still throwing so since previously the try/catch rethrew errors the behavior is the same.

Reply:
Yeah, it's a weird way to handle errors. But you have to remember that we're only doing this because Electron doesn't allow us to do this through normal error handling :d

I definitely don't think you should go ahead and add all FS errors. The idea is to only handle the errors that we care about through this system, and have some general error handling for all other cases. Actually now that I think about it, you don't need to use the try-catch here anymore, since the catch block would just rethrow the error.

So the only thing I would change here is to handle all potential errors we care about, e.g.
```ts
if (readDirResult.error) {
  if (readDirResult.code === FsErrorCode.FolderNotFound) {
    throw new Error('Folder not found')
  }
  throw new Error('Unexpected FsError: ' + readDirResult.error)
}
```

Reply:
ah sorry, my line comment here must have got swallowed at some point when I was making changes. 

This is throwing on all none missing file errors
https://github.com/ShieldBattery/ShieldBattery/pull/939/files#diff-be7de2630127e1ca9494b92af4ef574add9d864f56577fb428bef0b0d57fafbbR19

and this is setting the list of entries to be `[]` if there is no folder

https://github.com/ShieldBattery/ShieldBattery/pull/939/files#diff-be7de2630127e1ca9494b92af4ef574add9d864f56577fb428bef0b0d57fafbbR22

This explains why I was doing that https://github.com/ShieldBattery/ShieldBattery/issues/936. This is fine for the current usages of readFile, but i'm not sure if it's a safe thing to do in general

Reply:
Yeah, that doesn't sound like a great idea to do here. I'm not sure we want to return an empty array every time we try to read a folder that doesn't exist.

You could potentially extend this error you're throwing here with our custom error code so you can check it in place it's used, in case you want to handle that error specially at that place. But I'm not sure that #936 is even the best way to handle that case in the first place. If you're missing the root replay folder, shouldn't we let the users know that and maybe how they can fix that? I guess it's potentially fine doing what you suggested too, if the user doesn't really need to do anything to fix it since the folder will be created when they play their first game?

But in any case, I would probably add the error code to the error you're throwing here, so the file-browser can check that code and do whatever it wants when the folder doesn't exist.

Reply:
Alright I updated to have the same behavior as before.

> If you're missing the root replay folder, shouldn't we let the users know that and maybe how they can fix that? 

It doesn't exist for fresh installs. I happened to run into the error after setting up a new PC.